llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp

```diff
@@ -7726,3 +7726,65 @@ bool CombinerHelper::matchShuffleUndefRHS(MachineInstr &MI,
 
   return true;
 }
+
+static void commuteMask(MutableArrayRef<int> Mask, const unsigned NumElems) {
+  const unsigned MaskSize = Mask.size();
+  for (unsigned I = 0; I < MaskSize; ++I) {
+    int Idx = Mask[I];
+    if (Idx < 0)
+      continue;
+
+    if (Idx < (int)NumElems)
+      Mask[I] = Idx + NumElems;
+    else
+      Mask[I] = Idx - NumElems;
+  }
+}
+
+bool CombinerHelper::matchShuffleDisjointMask(MachineInstr &MI,
+                                              BuildFnTy &MatchInfo) {
+
+  auto &Shuffle = cast<GShuffleVector>(MI);
+  // If any of the two inputs is already undef, don't check the mask again to
+  // prevent infinite loop
+  if (getOpcodeDef(TargetOpcode::G_IMPLICIT_DEF, Shuffle.getSrc1Reg(), MRI))
+    return false;
+
+  if (getOpcodeDef(TargetOpcode::G_IMPLICIT_DEF, Shuffle.getSrc2Reg(), MRI))
+    return false;
+
+  ArrayRef<int> Mask = Shuffle.getMask();
+  const LLT Src1Ty = MRI.getType(Shuffle.getSrc1Reg());
+
+  const unsigned NumSrcElems = Src1Ty.isVector() ? Src1Ty.getNumElements() : 1;
+
+  bool TouchesSrc1 = false;
+  bool TouchesSrc2 = false;
+  const unsigned NumElems = Mask.size();
+  for (unsigned Idx = 0; Idx < NumElems; ++Idx) {
+    if (Mask[Idx] < 0)
+      continue;
+
+    if (Mask[Idx] < (int)NumSrcElems)
+      TouchesSrc1 = true;
+    else
+      TouchesSrc2 = true;
+  }
+
+  if (!(TouchesSrc1 ^ TouchesSrc2))
+    return false;
+
+  Register NewSrc1 = Shuffle.getSrc1Reg();
+  SmallVector<int, 16> NewMask(Mask);
+  if (TouchesSrc2) {
+    NewSrc1 = Shuffle.getSrc2Reg();
+    commuteMask(NewMask, NumSrcElems);
+  }
+
+  MatchInfo = [=, &Shuffle](MachineIRBuilder &B) {
+    Register Undef = B.buildUndef(Src1Ty).getReg(0);
```

Comment:
legality check for undef

Reply:
Maybe we should have a rule that undef is always legal for every type. I can't see why it would be any use to make it illegal.

Reply:
Nope. If you want to build something, you have to check for legality. There cannot be a curated list of opcodes with weaker rules.

Reply:
Please also check for the legality of the G_SHUFFLE_VECTOR.

Reply:
> Maybe we should have a rule that undef is always legal for every type. I can't see why it would be any use to make it illegal.

That might be a nice idea, and I was wondering the same thing for AArch64 recently. At least treating all "things of legal size" as legal implicitdef, like v2i16 that would not otherwise be legal but is the same as a i32.  We here be able to say that if a value already exists as the input to a combine then an implicit def of that type should always exist (or be legalized later). Checking the legality of the implicit def sounds like something that shouldn't even fail.

Reply:
GlobalISel is about weird devices. There are address spaces attached to a LLT. It must be a hard rule, if you built something you must check for legality first.

Reply:
I can add the legality check for G_IMPLICIT_DEF.

>Please also check for the legality of the G_SHUFFLE_VECTOR.

We are just replacing an operand here, types stay the same and shuffle masks are not part of the legality query, so nothing to check here?

Reply:
> Maybe we should have a rule that undef is always legal for every type. I can't see why it would be any use to make it illegal.

This rule needs to be more precise. Implicit def must only be legal for register types. That is, for any type that is legal in any operand of any operation, implicit_def must be legal for it.

Reply:
If you want to build something you **must** go through `isLegalOrBeforeLegalizer` .

Reply:
We don't have legality rules for COPY. COPY and [G_]IMPLICIT_DEF are a pair that at a low level need to be supported for any register value

Reply:
COPY might be different. But for G_IMPLCIT_DEF state for which types it is legal (maybe all) in your legalizer. Never ever assume that something is legal.

Reply:
> GlobalISel is about weird devices. 

To clarify for people reading this thread in future: no that's not specifically what we designed GlobalISel for. I'm not sure where this notion originated from.

> This rule needs to be more precise. Implicit def must only be legal for register types. That is, for any type that is legal in any operand of any operation, implicit_def must be legal for it.

+1

> I agree that is odd one. I saw something then I am allowed to build it again?

Yes. Whether or not something is legal is context independent. Note there's some subtlety here, you're allowed to make different optimization decisions in the legalizer based on context but the same instruction + type tuple cannot be illegal in one context and legal in another.

If you weren't allowed to assume that an existing tuple in the MIR implied it was ok to build another, then something simple like instruction duplication or even moving would be incorrect. It also prevents us from caching legality which we may want to do in future for compile time performance.

> If you want to build something you must go through isLegalOrBeforeLegalizer .
> COPY might be different. But for G_IMPLCIT_DEF state for which types it is legal (maybe all) in your legalizer. Never ever assume that something is legal

I'm not sure what makes you think that your opinion is the final word on any of this, but this isn't your decision to make unilaterally. Please cool it with this kind of language, I noted on other PRs like Craig Topper's that your wording is unnecessarily harsh. I see it again here.

Reply:
> COPY might be different. But for G_IMPLCIT_DEF state for which types it is legal (maybe all) in your legalizer. Never ever assume that something is legal.

And there are restrictions on what the legality rules must be. We should enforce in the verifier that G_IMPLICIT_DEF is legal for all types used in any operand, such that it should never be necessary to check if it is legal

Reply:
Sorry, for sounding harsh. I(me) find it confusing to be weak about legality. Sometimes you have to go through isLegalOrBeforeLegalizer but not always, because undef is different. Why limit it to undef?

Can we instead document guidelines for building legalizers and using/querying them.

Reply:
> Sometimes you have to go through isLegalOrBeforeLegalizer but not always, because undef is different. Why limit it to undef?

See:
> We don't have legality rules for COPY. COPY and [G_]IMPLICIT_DEF are a pair that at a low level need to be supported for any register value

... to elaborate further: if you can create a value with a type T, you always need to be able to create an IMPLICIT_DEF with that type. Likewise you also need to be able to COPY it (we could have introduced G_COPY and we'd be having the same discussion about that here). Implicit-def is just an expression of an undefined value of a certain type. The fact that you saw an instruction that uses or defines that type by definition means that you also should be able to have an implicit def of it.